[10.0][ADD] account_tax_change

sebalix opened the pull request:
Backport of:
- https://github.com/OCA/account-invoicing/pull/1433

sebalix committed:
account_tax_change: declare tax changes and update invoices

sebalix committed:
account_tax_change: backport to 10.0

grindtildeath commented on account_tax_change/models/account_tax_change.py:
```diff
@@ -0,0 +1,62 @@
+    type_tax_use = fields.Selection(
+        related="tax_src_id.type_tax_use",
+    )
+    tax_dest_id = fields.Many2one(
```
I guess we need a constraint to somehow make sure `tax_src_id` and `tax_dest_id` have the same `type_tax_use`.

grindtildeath commented on account_tax_change/security/ir.model.access.csv:
```diff
@@ -0,0 +1,9 @@
+access_account_tax_change_line_invoice,account.tax.change.line,model_account_tax_change_line,account.group_account_invoice,1,0,0,0
+access_account_tax_change_manager,account.tax.change,model_account_tax_change,account.group_account_manager,1,1,1,1
+access_account_tax_change_line_manager,account.tax.change.line,model_account_tax_change_line,account.group_account_manager,1,1,1,1
+access_account_move_apply_tax_change_invoice,account.move.apply.tax.change,model_account_move_apply_tax_change,account.group_account_invoice,1,1,1,1
+access_account_move_apply_tax_change_manager,account.move.apply.tax.change.manager,model_account_move_apply_tax_change,account.group_account_manager,1,1,1,1
```
IIRC there are no access rights for wizards in v10.0

grindtildeath commented on account_tax_change/wizards/account_move_apply_tax_change.py:
```diff
@@ -0,0 +1,95 @@
+    def validate(self):
+        """Apply the tax changes on the selected invoices."""
+        self.ensure_one()
+        src_taxes = self.tax_change_id.change_line_ids.tax_src_id
```
I guess we need a `mapped` here in v10

grindtildeath commented on account_tax_change/wizards/account_move_apply_tax_change.py:
```diff
@@ -0,0 +1,95 @@
+
+    def _skip_invoice(self, invoice):
+        """No tax change on invoice already posted or paid."""
+        return invoice.state == "posted"
```
```suggestion
        return invoice.state in ("open", "paid", "cancel")
```

sebalix replied:
We should skip `open` invoices in your opinion? In the `invoice_ids` M2M above I added a domain that allows them.
EDIT: And I added a `_check_invoices` used by `default_get` as well to check the invoices state, so this `skip_invoice` will do nothing now, but we could keep it as a hook.

grindtildeath replied:
Open invoices are the "posted" ones which did already generate an account.move, so in case we modify such an invoice, we'd have to either modify or regenerate the move. I guess it's then easier to restrict application to invoices that did not generate a move.

sebalix replied:
Yes I agree, it's easier for me ;) But was wondering if it was correct as it probably depends also on the localization.

grindtildeath commented on account_tax_change/wizards/account_move_apply_tax_change.py:
```diff
@@ -0,0 +1,95 @@
+        required=True,
+        string="Invoices",
+        domain=(
+            "[('state', 'in', ('draft', 'open')),"
```
I guess we could allow `proforma` and `proforma2` here (https://github.com/odoo/odoo/blob/10.0/addons/account/models/account_invoice.py#L228) :thinking:

grindtildeath commented on account_tax_change/wizards/account_move_apply_tax_change.py:
```diff
@@ -0,0 +1,95 @@
+        res = super(AccountMoveApplyTaxChange, self).default_get(fields_list)
+        active_model = self.env.context.get("active_model")
+        if active_model == "account.invoice":
+            invoice_ids = self.env.context.get("active_ids")
+            invoices = self.env[active_model].search(
+                [
+                    ("id", "in", invoice_ids),
+                    ("state", "in", ("draft", "open")),
+                ]
+            )
+            res["invoice_ids"] = [(6, 0, invoices.ids)]
```
IMO we should raise an error if one of the selected invoices is not in an "allowed" state.

grindtildeath commented on account_tax_change/wizards/account_move_apply_tax_change.py:
```diff
@@ -0,0 +1,95 @@
+        Other modules could override this method.
+        """
+        return (
+            not line.invoice_id.date_invoice
+            or line.invoice_id.date_invoice < self.tax_change_id.date
```
IMO this should also go in `_skip_invoice` as the date is not defined on the lines. (but it's good to keep the hook)

grindtildeath replied:
Actually thinking through this one is tricky because the application of tax change will depend if the product that is sold is supplied after the tax change date, and we cannot know this.
A same product could be:
 - sold in 2022 and supplied in 2022 (in such case no tax change is required) 
 - sold in 2022 and supplied in 2023 (in such case tax change is required)

IMO let's remove the condition on the invoice date, as the user has to know what he's doing.

sebalix replied:
I have some doubts about this, the tax to apply isn't based on the invoice date, whatever the shipping date is?

sebalix replied:
It depends... reading this for France: https://www.demarches.interieur.gouv.fr/professionnels/quelles-dates-exigibilite-matiere-tva
so you are right, better to remove this constraint and let the user decide.

grindtildeath reviewed (commented):
Few things to improve, and missing utf-8 headers for python2

sebalix commented:
@grindtildeath I addressed all your remarks, what was relevant for v16 has been done there first (https://github.com/OCA/account-invoicing/pull/1433) as fixups and backported here + adapted if needed in the last commit.

grindtildeath commented on account_tax_change/wizards/account_move_apply_tax_change.py:
```diff
@@ -0,0 +1,107 @@
+        return invoice.state in ("open", "paid", "cancel")
+
+    def _skip_line(self, line):
+        """Return `True` if the tax change should be applied on the line.
```
docstring to fix. Shouldn't we return false as default instead of an implicit None?

sebalix replied:
We could return False yes

grindtildeath reviewed (commented):
Just wondering, shouldn't we change the module name to be a bit more specific?
i.e. it's changing taxes on invoices and not on changing the tax themselves, so maybe something like account_invoice_tax_change?

grindtildeath commented on account_tax_change/wizards/account_move_apply_tax_change.py:
```diff
@@ -0,0 +1,108 @@
+
+        Other modules could override this method.
+        """
+        line.invoice_line_tax_ids -= self.tax_change_id.change_line_ids.tax_src_id
+        line.invoice_line_tax_ids |= self.tax_change_id.change_line_ids.tax_dest_id
```
```suggestion
        for tax_change_line in self.tax_change_id.change_line_ids:
            if tax_change_line.tax_src_id in line.invoice_line_tax_ids:
                line.invoice_line_tax_ids -= tax_change_line.tax_src_id
                line.invoice_line_tax_ids |= tax_change_line.tax_dest_id
```

sebalix replied:
fixed
